Core/Core/SwiftUIViews/RefreshableScrollView.swift

```diff
@@ -0,0 +1,141 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2022-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+public struct RefreshableScrollView<Content: View>: View {
+    enum ViewState {
+        case progress(CGFloat)
+        case animating
+    }
+
+    // MARK: - Dependencies
+
+    private let axes: Axis.Set
+    private let showsIndicator: Bool
+    private let content: () -> Content
+    private let refreshAction: (@escaping () -> Void) -> Void
+
+    // MARK: - Private properties
+
+    @State private var isAnimating = false
+    @State private var progress: CGFloat = 0
+    @State private var viewState: ViewState = .animating
+    @State private var offset: CGFloat = 0
+    private let snappingPoint: CGFloat = 64
+    private let hapticGenerator = UIImpactFeedbackGenerator(style: .light)
+
+    // MARK: - Init
+
+    public init(
+        _ axes: Axis.Set = .vertical,
+        showsIndicators: Bool = true,
+        @ViewBuilder content: @escaping () -> Content,
+        refreshAction: @escaping (@escaping () -> Void) -> Void
+    ) {
+        self.axes = axes
+        showsIndicator = showsIndicators
+        self.content = content
+        self.refreshAction = refreshAction
+    }
+
+    public var body: some View {
+        ScrollView(
+            axes,
+            showsIndicators: showsIndicator,
+            content: {
+                VStack(spacing: 0) {
+                    switch viewState {
+                    case .animating:
+                        indeterminateProgressView()
+                    case let .progress(progress):
+                        determinateProgressView(progress)
+                    }
+                    content()
+                        .offset(x: 0, y: isAnimating ? 0 : -48)
```

Comment:
We could extract these constants to be meaningful. I guess this 48 is the indeterminateCircle factory method's default 32 point size plus the 16 point padding above the progress circle.

Reply:
Changed!